cql3/restrictions/statement_restrictions.cc

```diff
@@ -1003,5 +1006,725 @@ ::shared_ptr<single_column_restriction> single_column_restriction::LIKE::apply_t
+};
+
+/// Returns col's value from queried data.
+bytes_opt get_value_from_partition_slice(
+        const column_value& col, row_data_from_partition_slice data, const query_options& options) {
+    auto cdef = col.col;
+    if (col.sub) {
+        auto col_type = static_pointer_cast<const collection_type_impl>(cdef->type);
+        if (!col_type->is_map()) {
+            throw exceptions::invalid_request_exception(format("subscripting non-map column {}", cdef->name_as_text()));
+        }
+        const auto deserialized = cdef->type->deserialize(*data.other_columns[data.sel.index_of(*cdef)]);
+        const auto& data_map = value_cast<map_type_impl::native_type>(deserialized);
+        const auto key = col.sub->bind_and_get(options);
+        auto&& key_type = col_type->name_comparator();
+        const auto found = with_linearized(*key, [&] (bytes_view key_bv) {
+            using entry = std::pair<data_value, data_value>;
+            return std::find_if(data_map.cbegin(), data_map.cend(), [&] (const entry& element) {
+                return key_type->compare(element.first.serialize_nonnull(), key_bv) == 0;
+            });
+        });
+        return found == data_map.cend() ? bytes_opt() : bytes_opt(found->second.serialize_nonnull());
+    } else {
+        switch (cdef->kind) {
+        case column_kind::partition_key:
+            return data.partition_key[cdef->id];
+        case column_kind::clustering_key:
+            return data.clustering_key[cdef->id];
+        case column_kind::static_column:
+        case column_kind::regular_column:
+            return data.other_columns[data.sel.index_of(*cdef)];
+        default:
+            throw exceptions::unsupported_operation_exception("Unknown column kind");
+        }
+    }
+}
+
+/// Returns col's value from a mutation.
+bytes_opt get_value_from_mutation(const column_value& col, row_data_from_mutation data) {
+    const auto v = do_get_value(
+            data.schema_, *col.col, data.partition_key_, data.clustering_key_, data.other_columns, data.now);
+    return v ? v->linearize() : bytes_opt();
+}
+
+/// Returns col's value from the fetched data.
+bytes_opt get_value(const column_value& col, column_value_eval_bag bag) {
+    using std::placeholders::_1;
+    return std::visit(overloaded_functor{
+            std::bind(get_value_from_mutation, col, _1),
+            std::bind(get_value_from_partition_slice, col, _1, bag.options),
+        }, bag.row_data);
+}
+
+/// Type for comparing results of get_value().
+const abstract_type* get_value_comparator(const column_value& cv) {
+    auto col_type = cv.col->type;
+    if (cv.sub) {
+        return static_pointer_cast<const collection_type_impl>(col_type)->value_comparator().get();
+    }
+    return col_type->is_reversed() ? col_type->underlying_type().get() : col_type.get();
+}
+
+/// Returns a tuple-valued terminal from t, if possible.  Otherwise, returns null.
```

Comment:
Because a term can be a non-tuple non_terminal.  I'll try to clarify in the blurb.

Reply:
Of course a term can be a non-tuple, but in general the grammar is type-safe and strict. Any bind marker has just one type. So we should know statically if it's a tuple or not.

Reply:
> So we should know statically if it's a tuple or not.

Does that mean the binary-operator's RHS should be changed from generic `term` to a variant of many possible `term` subtypes?  I like that direction, though I was hoping to defer it into a future rewriting step (that also drops the term hierarchy).

Reply:
Probably we should (and not now), but the question I'm trying to answer is "does returning nullptr here indicate a scylla bug, or a user error, or something else?"

Reply:
> "does returning nullptr here indicate a scylla bug, or a user error, or something else?"

Something else -- it means `t` is a non-terminal that cannot be evaluated to a tuple.  The caller then decides if this is an error (eg, because LHS is a tuple) or a case where `t` has a single item that should be obtained via bind_and_get(), rather than get_elements().

Reply:
Can you give an example for the second case?

Reply:
It's actually right [here](https://github.com/scylladb/scylla/pull/5763/files/9fd7e3ce646493c7ba0f6a70a4553c4d9c3ac164..4b4432873f9f013344cd20a3f08460d3d35a5561#diff-9753ff707c1cb3d126e0c90f8486f5b2R1229) (scroll to the bottom, click "Load diff", sigh...).  See how it calls get_elements() for a tuple but bind_and_get() for a non-tuple?

Reply:
I see. Is it correct that if we had a tuple_equal operator then the two cases would be disambiguated? I'm trying to increase my understanding now, not suggesting a change yet.

In general everything should be statically types and we should never have to examine the dynamic type. e.g. we should be able to replace all dynamic_casts with static_casts.

Reply:
> Is it correct that if we had a tuple_equal operator then the two cases would be disambiguated?

Yes.  If we knew whether the LHS was a single-element tuple or just a column, we wouldn't need to check the type of the RHS.  We'd simply invoke get_tuple()->get_elements() or bind_and_get(), respectively, to obtain the RHS value for comparison.

> In general everything should be statically types and we should never have to examine the dynamic type. e.g. we should be able to replace all dynamic_casts with static_casts.

You're right.  In fact, I believe I can statically cast every get_tuple() result to tuples::value even now.  I'll do that in the next version.

Reply:
Ok. So please add comments explaining the reasons for the remaining dynamic_casts, and we'll later update the type system so we can remove them.

I checked, and ubsan warns on invalid static_casts, so we'll still get checks in debug mode.

Reply:
Simplified code and improved the blurb.